client/src/components/DependentCompositeField/DependentCompositeField.js

```diff
@@ -0,0 +1,197 @@
+  actions,
+  // initialChildrenSchema represents the children schema prior to calling any dependency callbacks
+  // and should not be used directly when building child components.
+  childrenSchema: initialChildrenSchema,
+  className = '',
+  data,
+  extraClass = '',
+  formFieldSchemaFunctions,
+  formid,
+  ...props
+}) => {
+  const Tag = data.tag || 'div';
+  // childrenSchema reprensents the children schema after calling any dependency callbacks
+  // and may not always match what is passed in through props.
+  const [childrenSchema, setChildrenSchema] = useState(initialChildrenSchema);
+  const [loading, setLoading] = useState(false);
+  const reduxState = useSelector(storedState => storedState);
+
+  // This effect fires any time a child field message changes, e.g. for validation messages.
+  // It adds messages from the prop schema into the state schema.
+  // This is necessary because FormBuilderLoader.reduceSchemaErrors updates the state which will
+  // be reflected in the prop value - but that prop may have outdated information in other respects
+  // such as which component to render.
+  useEffect(() => {
+    // If we have no schema prop, there's no change to make.
+    if (!initialChildrenSchema) {
+      return;
+    }
+    // The logic is done in a callback so that if this happens in the same
+    // cycle as the call in the change handler, we update that state rather
+    // than replacing it.
+    setChildrenSchema((oldSchema) => {
+      const newSchema = [];
+      let hasChange = false;
+      for (const initialChild of initialChildrenSchema) {
+        const newChild = Object.assign({}, oldSchema.find((item) => item.id === initialChild.id));
+        // If the new message is different to what we have, grab the message.
+        if (newChild.message?.rawValue !== initialChild.message?.rawValue) {
+          newChild.message = initialChild.message;
+          hasChange = true;
+        }
+        newSchema.push(newChild);
+      }
+      // Return the original state if no changes were made.
+      // React won't trigger a rerender if it detects the state didn't change.
+      return hasChange ? newSchema : oldSchema;
+    });
+  }, [...initialChildrenSchema?.map(childField => [childField.name, childField.message?.rawValue]) ?? []]);
+
+  /**
+   * Handle changes to internal field values and update the schema appropriately.
+   * Using useCallback and debounce to prevent issues fetching a new schema while the user is still typing.
+   */
+  const handleChange = useCallback(debounce((value, state, schema, allChildrenNames) => {
+    setLoading(true);
+    // Call existing callback if there was one
+    if (typeof schema.onChange === 'function') {
+      schema.onChange(value);
+    }
+    // Get values from redux store but only for the relevant fields
+    const allValues = getFormValues(formid, getFormState)(state);
+    const onlyNecessaryValues = Object.fromEntries(Object.entries(allValues).filter(
```

Comment:
Need a comment explaining what this doing

Reply:
Have updated the comment to give a bit more detail but unless you tell me more specifically what's not clear about it I won't know what needs to be done here.